devlib/trace/dmesg.py

```diff
@@ -0,0 +1,185 @@
+#    Copyright 2019 ARM Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import division
+import re
+from itertools import takewhile
+from datetime import timedelta
+
+from devlib.trace import TraceCollector
+
+
+class KernelLogEntry(object):
+    """
+    Entry of the kernel ring buffer.
+
+    :param facility: facility the entry comes from
+    :type facility: str
+
+    :param level: log level
+    :type level: str
+
+    :param ts: Timestamp of the entry
+    :type ts: datetime.timedelta
+
+    :param msg: Content of the entry
+    :type msg: str
+    """
+    def __init__(self, facility, level, ts, msg):
+        self.facility = facility
+        self.level = level
+        self.ts = ts
+        self.msg = msg
+
+    @classmethod
+    def from_str(cls, line):
+        """
+        Parses a "dmesg --decode" output line, formatted as following:
+        kern  :err   : [3618282.310743] nouveau 0000:01:00.0: systemd-logind[988]: nv50cal_space: -16
+
+        Or the more basic output:
+        [3618282.310743] nouveau 0000:01:00.0: systemd-logind[988]: nv50cal_space: -16
+
+        """
+
+        def parse_ts_msg(line):
+            ts_msg_regex = r'\[(.*?)\] (.*)'
+            match = re.match(ts_msg_regex, line.strip())
+            if not match:
+                raise ValueError('dmesg entry format not recognized: {}'.format(line))
+            ts, msg = match.groups()
+            return ts, msg
+
+        # If we can parse the timestamp directly, that is a basic line
+        try:
+            ts, msg = parse_ts_msg(line)
+        except ValueError:
+            facility, level, remainder = line.split(':', 2)
+            ts, msg = parse_ts_msg(remainder)
```

Comment:
Wrap this in a try catch as well in case the second parsing attempt also fails?

Reply:
If this one fails, that means the parser is broken so we want to know about it to fix it. The 1st failure is somehow expected and should always detect correctly what it is supposed to detect (i.e. something before the timestamp `[42.4242]` or not).